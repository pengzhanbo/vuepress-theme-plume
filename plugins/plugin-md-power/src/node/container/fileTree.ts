import type { Markdown } from 'vuepress/markdown'
import type { FileTreeIconMode, FileTreeOptions } from '../../shared/index.js'
import Token from 'markdown-it/lib/token.mjs'
import container from 'markdown-it-container'
import { removeEndingSlash, removeLeadingSlash } from 'vuepress/shared'
import { defaultFile, defaultFolder, getFileIcon } from '../fileIcons/index.js'
import { resolveAttrs } from '../utils/resolveAttrs.js'

interface FileTreeNode {
  filename: string
  type: 'folder' | 'file'
  expanded: boolean
  focus: boolean
  empty: boolean
}

interface FileTreeAttrs {
  title?: string
  icon?: FileTreeIconMode
}

const type = 'file-tree'
const closeType = `container_${type}_close`
const componentName = 'FileTreeItem'
const itemOpen = 'file_tree_item_open'
const itemClose = 'file_tree_item_close'

export function fileTreePlugin(md: Markdown, options: FileTreeOptions = {}) {
  const getIcon = (filename: string, type: 'folder' | 'file', mode?: FileTreeIconMode): string => {
    mode ||= options.icon || 'colored'
    if (mode === 'simple')
      return type === 'folder' ? defaultFolder : defaultFile
    return getFileIcon(filename, type)
  }

  const validate = (info: string): boolean => info.trim().startsWith(type)

  const render = (tokens: Token[], idx: number): string => {
    const { attrs } = resolveAttrs<FileTreeAttrs>(tokens[idx].info.slice(type.length - 1))

    if (tokens[idx].nesting === 1) {
      const hasRes: number[] = [] // level stack
      for (
        let i = idx + 1;
        !(tokens[i].nesting === -1
          && tokens[i].type === closeType);
        ++i
      ) {
        const token = tokens[i]
        if (token.type === 'list_item_open') {
          const result = resolveTreeNodeInfo(tokens, token, i)
          if (result) {
            hasRes.push(token.level)
            const [info, inline] = result
            const { filename, type, expanded, empty } = info
            const icon = getIcon(filename, type, attrs.icon)

            token.type = itemOpen
            token.tag = componentName
            token.attrSet('type', type)
            token.attrSet(':expanded', expanded ? 'true' : 'false')
            token.attrSet(':empty', empty ? 'true' : 'false')
            updateInlineToken(inline, info, icon)
          }
          else {
            hasRes.push(-1)
          }
        }
        else if (token.type === 'list_item_close') {
          if (token.level === hasRes.pop()) {
            token.type = itemClose
            token.tag = componentName
          }
        }
      }
      const title = attrs.title
      return `<div class="vp-file-tree">${title ? `<p class="vp-file-tree-title">${title}</p>` : ''}`
    }
    else {
      return '</div>'
    }
  }

  md.use(container, type, { validate, render })
}

export function resolveTreeNodeInfo(
  tokens: Token[],
  current: Token,
  idx: number,
): [FileTreeNode, Token] | undefined {
  let hasInline = false
  let hasChildren = false
  let inline!: Token
  for (
    let i = idx + 1;
    !(tokens[i].level === current.level && tokens[i].type === 'list_item_close');
    ++i
  ) {
    if (tokens[i].type === 'inline' && !hasInline) {
      inline = tokens[i]
      hasInline = true
    }
    else if (tokens[i].tag === 'ul') {
      hasChildren = true
    }

    if (hasInline && hasChildren)
      break
  }

  if (!hasInline)
    return undefined

  const children = inline.children?.filter(token => (token.type === 'text' && token.content) || token.tag === 'strong') || []
  const filename = children.filter(token => token.type === 'text').map(token => token.content).join(' ').split(/\s+/)[0] ?? ''
  const focus = children[0]?.tag === 'strong'
  const type = hasChildren || filename.endsWith('/') ? 'folder' : 'file'
  const info: FileTreeNode = {
    filename: removeLeadingSlash(removeEndingSlash(filename)),
    type,
    focus,
    empty: !hasChildren,
    expanded: type === 'folder' && !filename.endsWith('/'),
  }

  return [info, inline] as const
}

export function updateInlineToken(inline: Token, info: FileTreeNode, icon: string) {
  const children = inline.children
  if (!children)
    return

  const tokens: Token[] = []
  const wrapperOpen = new Token('span_open', 'span', 1)
  const wrapperClose = new Token('span_close', 'span', -1)

  wrapperOpen.attrSet('class', `tree-node ${info.type}`)
  tokens.push(wrapperOpen)

  if (info.filename !== '...' && info.filename !== 'â€¦') {
    const iconOpen = new Token('vp_iconify_open', 'VPIcon', 1)
    iconOpen.attrSet('name', icon)
    const iconClose = new Token('vp_iconify_close', 'VPIcon', -1)

    tokens.push(iconOpen, iconClose)
  }

  const fileOpen = new Token('span_open', 'span', 1)
  fileOpen.attrSet('class', `name${info.focus ? ' focus' : ''}`)
  tokens.push(fileOpen)

  let isStrongTag = false
  while (children.length) {
    const token = children.shift()!
    if (token.type === 'text' && token.content) {
      if (token.content.includes(' ')) {
        const [first, ...other] = token.content.split(' ')
        const text = new Token('text', '', 0)
        text.content = removeEndingSlash(first)
        tokens.push(text)
        const comment = new Token('text', '', 0)
        comment.content = other.join(' ')
        children.unshift(comment)
      }
      else {
        token.content = removeEndingSlash(token.content)
        tokens.push(token)
      }
      if (!isStrongTag)
        break
    }
    else if (token.tag === 'strong') {
      token.content = removeEndingSlash(token.content)
      tokens.push(token)
      if (token.nesting === 1) {
        isStrongTag = true
      }
      else {
        break
      }
    }
    else {
      tokens.push(token)
    }
  }

  const fileClose = new Token('span_close', 'span', -1)
  tokens.push(fileClose)

  if (children.filter(token => token.type === 'text' && token.content.trim()).length) {
    const commentOpen = new Token('span_open', 'span', 1)
    commentOpen.attrSet('class', 'comment')
    const commentClose = new Token('span_close', 'span', -1)

    tokens.push(commentOpen, ...children, commentClose)
  }

  tokens.push(wrapperClose)
  inline.children = tokens
}
