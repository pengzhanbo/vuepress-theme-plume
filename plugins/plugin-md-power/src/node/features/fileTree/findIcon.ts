import { defaultFile, defaultFolder, definitions } from './definitions.js'

export function getFileIcon(fileName: string, type: 'file' | 'folder' = 'file'): string {
  const name = getFileIconName(fileName, type)
  if (!name)
    return type === 'file' ? defaultFile : defaultFolder

  return name
}

function getFileIconName(fileName: string, type: 'file' | 'folder' = 'file'): string | undefined {
  if (type === 'folder') {
    return definitions.folders[fileName]
  }
  let icon: string | undefined = definitions.named[fileName] || definitions.files[fileName]
  if (icon)
    return icon
  icon = getFileIconTypeFromExtension(fileName)
  if (icon)
    return icon
  for (const [partial, partialIcon] of Object.entries(definitions.partials)) {
    if (fileName.includes(partial))
      return partialIcon
  }
  return icon
}

function getFileIconTypeFromExtension(fileName: string): string | undefined {
  const firstDotIndex = fileName.indexOf('.')
  if (firstDotIndex === -1)
    return
  let extension = fileName.slice(firstDotIndex)
  while (extension !== '') {
    const icon = definitions.extensions[extension]
    if (icon)
      return icon
    const nextDotIndex = extension.indexOf('.', 1)
    if (nextDotIndex === -1)
      return
    extension = extension.slice(nextDotIndex)
  }
  return undefined
}
