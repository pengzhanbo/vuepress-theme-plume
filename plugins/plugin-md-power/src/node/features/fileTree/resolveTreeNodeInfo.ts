import Token from 'markdown-it/lib/token.mjs'
import { removeEndingSlash, removeLeadingSlash } from 'vuepress/shared'

interface FileTreeNode {
  filename: string
  type: 'folder' | 'file'
  expanded: boolean
  focus: boolean
  empty: boolean
}

export function resolveTreeNodeInfo(
  tokens: Token[],
  current: Token,
  idx: number,
): [FileTreeNode, Token] | undefined {
  let hasInline = false
  let hasChildren = false
  let inline!: Token
  for (
    let i = idx + 1;
    !(tokens[i].level === current.level && tokens[i].type === 'list_item_close');
    ++i
  ) {
    if (tokens[i].type === 'inline' && !hasInline) {
      inline = tokens[i]
      hasInline = true
    }
    else if (tokens[i].tag === 'ul') {
      hasChildren = true
    }

    if (hasInline && hasChildren)
      break
  }

  if (!hasInline)
    return undefined

  const children = inline.children?.filter(token => (token.type === 'text' && token.content) || token.tag === 'strong') || []
  const filename = children.filter(token => token.type === 'text').map(token => token.content).join(' ').split(/\s+/)[0] ?? ''
  const focus = children[0]?.tag === 'strong'
  const type = hasChildren || filename.endsWith('/') ? 'folder' : 'file'
  const info: FileTreeNode = {
    filename: removeLeadingSlash(removeEndingSlash(filename)),
    type,
    focus,
    empty: !hasChildren,
    expanded: type === 'folder' && !filename.endsWith('/'),
  }

  return [info, inline] as const
}

export function updateInlineToken(inline: Token, info: FileTreeNode, icon: string) {
  const children = inline.children
  if (!children)
    return

  const tokens: Token[] = []
  const wrapperOpen = new Token('span_open', 'span', 1)
  const wrapperClose = new Token('span_close', 'span', -1)

  wrapperOpen.attrSet('class', `tree-node ${info.type}`)
  tokens.push(wrapperOpen)

  if (info.filename !== '...' && info.filename !== 'â€¦') {
    const iconOpen = new Token('vp_iconify_open', 'VPIcon', 1)
    iconOpen.attrSet('name', icon)
    const iconClose = new Token('vp_iconify_close', 'VPIcon', -1)

    tokens.push(iconOpen, iconClose)
  }

  const fileOpen = new Token('span_open', 'span', 1)
  fileOpen.attrSet('class', `name${info.focus ? ' focus' : ''}`)
  tokens.push(fileOpen)

  let isStrongTag = false
  while (children.length) {
    const token = children.shift()!
    if (token.type === 'text' && token.content) {
      if (token.content.includes(' ')) {
        const [first, ...other] = token.content.split(' ')
        const text = new Token('text', '', 0)
        text.content = first
        tokens.push(text)
        const comment = new Token('text', '', 0)
        comment.content = other.join(' ')
        children.unshift(comment)
      }
      else {
        tokens.push(token)
      }
      if (!isStrongTag)
        break
    }
    else if (token.tag === 'strong') {
      tokens.push(token)
      if (token.nesting === 1) {
        isStrongTag = true
      }
      else {
        break
      }
    }
    else {
      tokens.push(token)
    }
  }

  const fileClose = new Token('span_close', 'span', -1)
  tokens.push(fileClose)

  if (children.filter(token => token.type === 'text' && token.content.trim()).length) {
    const commentOpen = new Token('span_open', 'span', 1)
    commentOpen.attrSet('class', 'comment')
    const commentClose = new Token('span_close', 'span', -1)

    tokens.push(commentOpen, ...children, commentClose)
  }

  tokens.push(wrapperClose)
  inline.children = tokens
}
