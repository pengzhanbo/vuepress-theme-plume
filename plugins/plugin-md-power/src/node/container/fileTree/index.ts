import type Token from 'markdown-it/lib/token.mjs'
import type { Markdown } from 'vuepress/markdown'
import container from 'markdown-it-container'
import { getFileIcon } from './findIcon.js'
import { resolveTreeNodeInfo, updateInlineToken } from './resolveTreeNodeInfo.js'

const type = 'file-tree'
const closeType = `container_${type}_close`
const componentName = 'FileTreeItem'
const itemOpen = 'file_tree_item_open'
const itemClose = 'file_tree_item_close'

export function fileTreePlugin(md: Markdown) {
  const validate = (info: string): boolean => info.trim().startsWith(type)
  const render = (tokens: Token[], idx: number): string => {
    if (tokens[idx].nesting === 1) {
      const hasRes: number[] = [] // level stack
      for (
        let i = idx + 1;
        !(tokens[i].nesting === -1
          && tokens[i].type === closeType);
        ++i
      ) {
        const token = tokens[i]
        if (token.type === 'list_item_open') {
          const result = resolveTreeNodeInfo(tokens, token, i)
          if (result) {
            hasRes.push(token.level)
            const [info, inline] = result
            const { filename, type, expanded, empty } = info
            const icon = getFileIcon(filename, type)

            token.type = itemOpen
            token.tag = componentName
            token.attrSet('type', type)
            token.attrSet(':expanded', expanded ? 'true' : 'false')
            token.attrSet(':empty', empty ? 'true' : 'false')
            updateInlineToken(inline, info, icon)
          }
          else {
            hasRes.push(-1)
          }
        }
        else if (token.type === 'list_item_close') {
          if (token.level === hasRes.pop()) {
            token.type = itemClose
            token.tag = componentName
          }
        }
      }
      const info = tokens[idx].info.trim()

      const title = info.slice(type.length).trim()
      return `<div class="vp-file-tree">${title ? `<p class="vp-file-tree-title">${title}</p>` : ''}`
    }
    else {
      return '</div>'
    }
  }

  md.use(container, type, { validate, render })
}
