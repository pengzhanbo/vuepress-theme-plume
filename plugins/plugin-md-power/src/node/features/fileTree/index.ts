import fs from 'node:fs'
import container from 'markdown-it-container'
import type Token from 'markdown-it/lib/token.mjs'
import type { App } from 'vuepress/core'
import type { Markdown } from 'vuepress/markdown'
import { type FileIcon, folderIcon, getFileIcon } from './findIcon.js'
import { resolveTreeNodeInfo, updateInlineToken } from './resolveTreeNodeInfo.js'

const type = 'file-tree'
const closeType = `container_${type}_close`
const componentName = 'FileTreeItem'
const itemOpen = 'file_tree_item_open'
const itemClose = 'file_tree_item_close'
const classPrefix = 'vp-fti-'
const styleFilepath = 'internal/md-power/file-tree.css'

export async function fileTreePlugin(app: App, md: Markdown) {
  const validate = (info: string): boolean => info.trim().startsWith(type)
  const render = (tokens: Token[], idx: number): string => {
    if (tokens[idx].nesting === 1) {
      const hasRes: number[] = [] // level stack
      for (
        let i = idx + 1;
        !(tokens[i].nesting === -1
          && tokens[i].type === closeType);
        ++i
      ) {
        const token = tokens[i]
        if (token.type === 'list_item_open') {
          const result = resolveTreeNodeInfo(tokens, token, i)
          if (result) {
            hasRes.push(token.level)
            const [info, inline] = result
            const { filename, type, expanded, empty } = info
            const icon = type === 'file' ? getFileIcon(filename) : folderIcon

            token.type = itemOpen
            token.tag = componentName
            token.attrSet('type', type)
            token.attrSet(':expanded', expanded ? 'true' : 'false')
            token.attrSet(':empty', empty ? 'true' : 'false')
            updateInlineToken(inline, info, `${classPrefix}${icon.name}`)
            addIcon(icon)
          }
          else {
            hasRes.push(-1)
          }
        }
        else if (token.type === 'list_item_close') {
          if (token.level === hasRes.pop()) {
            token.type = itemClose
            token.tag = componentName
          }
        }
      }
      const info = tokens[idx].info.trim()

      const title = info.slice(type.length).trim()
      return `<div class="vp-file-tree">${title ? `<p class="vp-file-tree-title">${title}</p>` : ''}`
    }
    else {
      return '</div>'
    }
  }

  let timer: NodeJS.Timeout | null = null
  const icons: Record<string, FileIcon> = {}

  function addIcon(icon: FileIcon) {
    icons[icon.name] = icon

    if (timer)
      clearTimeout(timer)
    timer = setTimeout(async () => {
      let content = ''
      for (const icon of Object.values(icons)) {
        content += `.${classPrefix}${icon.name} { --icon: ${icon.svg}; }\n`
      }
      await app.writeTemp(styleFilepath, content)
    }, 150)
  }

  md.use(container, type, { validate, render })

  if (!fs.existsSync(app.dir.temp(styleFilepath)))
    await app.writeTemp(styleFilepath, '')
}
