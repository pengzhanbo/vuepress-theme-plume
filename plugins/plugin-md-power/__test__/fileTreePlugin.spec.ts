import type { FileTreeOptions } from '../src/shared/fileTree.js'
import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { fileTreePlugin } from '../src/node/container/fileTree.js'

function createMarkdown(options?: FileTreeOptions) {
  return new MarkdownIt().use(fileTreePlugin, options)
}

describe('fileTreePlugin', () => {
  const code = `\
:::file-tree
- docs
  - README.md
  - foo.md
- src
  - client
    - components
      - **Navbar.vue**
    - index.ts \# comment
  - node
    - index.ts
- .gitignore
- package.json
:::

::: file-tree title="files"
- src
  - js
    - â€¦
  - vue/
  - css/
- README.md
:::

::: file-tree icon="simple"
- docs
- src
  - a.js
  - b.ts
- README.md
:::

::: file-tree
-
-
  -
:::

::: file-tree
:::
`
  it('should work with default options', () => {
    const md = createMarkdown()

    expect(md.render(code)).toMatchSnapshot()
  })
})
