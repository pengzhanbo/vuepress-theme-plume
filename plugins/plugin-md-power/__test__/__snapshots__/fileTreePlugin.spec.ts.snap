// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fileTreePlugin > should work with default options 1`] = `
"<div class="vp-file-tree"><ul>
<FileTreeItem type="folder" :expanded="true" :empty="false"><span class="tree-node folder"><VPIcon name="vscode-icons:folder-type-docs"></VPIcon><span class="name">docs</span></span>
<ul>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="flat-color-icons:info"></VPIcon><span class="name">README.md</span></span></FileTreeItem>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="vscode-icons:file-type-markdown"></VPIcon><span class="name">foo.md</span></span></FileTreeItem>
</ul>
</FileTreeItem>
<FileTreeItem type="folder" :expanded="true" :empty="false"><span class="tree-node folder"><VPIcon name="vscode-icons:folder-type-src"></VPIcon><span class="name">src</span></span>
<ul>
<FileTreeItem type="folder" :expanded="true" :empty="false"><span class="tree-node folder"><VPIcon name="vscode-icons:folder-type-client"></VPIcon><span class="name">client</span></span>
<ul>
<FileTreeItem type="folder" :expanded="true" :empty="false"><span class="tree-node folder"><VPIcon name="vscode-icons:folder-type-component"></VPIcon><span class="name">components</span></span>
<ul>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="vscode-icons:file-type-vue"></VPIcon><span class="name focus"><strong>Navbar.vue</strong></span></span></FileTreeItem>
</ul>
</FileTreeItem>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="vscode-icons:file-type-typescript"></VPIcon><span class="name">index.ts</span><span class="comment"># comment</span></span></FileTreeItem>
</ul>
</FileTreeItem>
<FileTreeItem type="folder" :expanded="true" :empty="false"><span class="tree-node folder"><VPIcon name="vscode-icons:default-folder"></VPIcon><span class="name">node</span></span>
<ul>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="vscode-icons:file-type-typescript"></VPIcon><span class="name">index.ts</span></span></FileTreeItem>
</ul>
</FileTreeItem>
</ul>
</FileTreeItem>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="vscode-icons:file-type-git"></VPIcon><span class="name">.gitignore</span></span></FileTreeItem>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="vscode-icons:file-type-node"></VPIcon><span class="name">package.json</span></span></FileTreeItem>
</ul>
</div><div class="vp-file-tree"><p class="vp-file-tree-title">files</p><ul>
<FileTreeItem type="folder" :expanded="true" :empty="false"><span class="tree-node folder"><VPIcon name="vscode-icons:folder-type-src"></VPIcon><span class="name">src</span></span>
<ul>
<FileTreeItem type="folder" :expanded="true" :empty="false"><span class="tree-node folder"><VPIcon name="vscode-icons:default-folder"></VPIcon><span class="name">js</span></span>
<ul>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><span class="name">â€¦</span></span></FileTreeItem>
</ul>
</FileTreeItem>
<FileTreeItem type="folder" :expanded="false" :empty="true"><span class="tree-node folder"><VPIcon name="vscode-icons:default-folder"></VPIcon><span class="name">vue</span></span></FileTreeItem>
<FileTreeItem type="folder" :expanded="false" :empty="true"><span class="tree-node folder"><VPIcon name="vscode-icons:folder-type-css"></VPIcon><span class="name">css</span></span></FileTreeItem>
</ul>
</FileTreeItem>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="flat-color-icons:info"></VPIcon><span class="name">README.md</span></span></FileTreeItem>
</ul>
</div><div class="vp-file-tree"><ul>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="vscode-icons:default-file"></VPIcon><span class="name">docs</span></span></FileTreeItem>
<FileTreeItem type="folder" :expanded="true" :empty="false"><span class="tree-node folder"><VPIcon name="vscode-icons:default-folder"></VPIcon><span class="name">src</span></span>
<ul>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="vscode-icons:default-file"></VPIcon><span class="name">a.js</span></span></FileTreeItem>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="vscode-icons:default-file"></VPIcon><span class="name">b.ts</span></span></FileTreeItem>
</ul>
</FileTreeItem>
<FileTreeItem type="file" :expanded="false" :empty="true"><span class="tree-node file"><VPIcon name="vscode-icons:default-file"></VPIcon><span class="name">README.md</span></span></FileTreeItem>
</ul>
</div><div class="vp-file-tree"><ul>
<li></li>
<li>
<ul>
<li></li>
</ul>
</li>
</ul>
</div><div class="vp-file-tree"></div>"
`;
